src/t_hash.c

```diff
@@ -185,6 +186,231 @@ static void hashDictWithExpireOnRelease(dict *d) {
     ebDestroy(&dictExpireMeta->hfe, &hashFieldExpireBucketsType, NULL);
 }
 
+#define HASH_LP_NO_TTL 0
+
+typedef struct listpackHfe {
+    ExpireMeta meta;
+    void *lp;
+    sds key;
+} listpackHfe;
+
+int hashLpIsHfe(void *ptr) { return (int) (((uintptr_t) ptr) & 1U); }
+void hashLpSetHfe(void **ptr) { *ptr = (void*) (((uintptr_t) *ptr) | 1); }
+listpackHfe *hashLpGet(void *ptr) {return (void*) ((uintptr_t) ptr & ~1);}
+void *hashLpGetListpack(void *ptr){return hashLpIsHfe(ptr) ? hashLpGet(ptr)->lp : ptr; }
+void hashLpSetListpack(robj *o, unsigned char *lp) {
+    if (hashLpIsHfe(o->ptr))
+        hashLpGet(o->ptr)->lp = lp;
+    else
+        o->ptr = lp;
+}
+
+static uint64_t hashLpExpireDryRun(const robj *o, uint64_t now) {
+    unsigned char *fptr;
+    unsigned char *lp = hashLpGetListpack(o->ptr);
+    uint64_t expired = 0;
+
+    fptr = lpFirst(lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lp, fptr);
+        fptr = lpNext(lp, fptr);
+        lpGetValue(fptr, NULL, &val);
+
+        if (val != 0 && val <= (long long) now)
+            expired++;
+
+        fptr = lpNext(lp, fptr);
+    }
+
+    return expired;
+}
+
+static void hashLpExpire(robj *o, ExpireInfo *info) {
+    long long min = LLONG_MAX;
+    unsigned char *fptr, *head;
+    listpackHfe *hfe = hashLpGet(o->ptr);
+
+    fptr = lpFirst(hfe->lp);
+    while (fptr != NULL) {
+        long long val;
+
+        head = fptr;
+        fptr = lpNext(hfe->lp, fptr);
+        fptr = lpNext(hfe->lp, fptr);
+
+        lpGetValue(fptr, NULL, &val);
+
+        if (val != 0 && val <= (long long) info->now) {
+            server.stat_expiredHashFields++;
+            hfe->lp = lpDeleteRangeWithEntry(hfe->lp, &head, 3);
+            fptr = head;
+            info->itemsExpired++;
+        } else {
+            if (val != 0)
+                min = min < val ? min : val;
+
+            fptr = lpNext(hfe->lp, fptr);
+        }
+    }
+
+    info->nextExpireTime = (min == LLONG_MAX) ? 0 : min;
+}
+
+/* Convert listpack to hold TTLs. New listpack will hold field name, value and
+ * ttl per field. */
+static void hashLpConvert(robj *obj, sds key) {
+    void *fptr;
+    long long val;
+    unsigned int slen;
+    listpackHfe *hfe = zmalloc(sizeof(*hfe));
+
+    hfe->lp = lpNew(lpBytes(obj->ptr) + lpLength(obj->ptr));
```

Comment:
one TTL field at least consumes 2 bytes, and let's add comment for it.
```suggestion
    hfe->lp = lpNew(lpBytes(obj->ptr) + lpLength(obj->ptr)*2);
```